### PR TITLE
Bring back old Hard-Coded Registry

### PR DESCRIPTION
**Pull request by yshrsmz**
#131 

brought back https://github.com/PointyCastle/pointycastle/commit/2da75e5a8d7bdbf95d08329add9f13b9070b75d4, which you need to call `initCipher` beforehand.

I'm currently working on creating a bit more dynamic registry as stated in here: https://github.com/PointyCastle/pointycastle/issues/131#issuecomment-412473515. I think it may take some time.

but anyway, here's the current status.

### CODE REVIEW

File: lib/asymmetric/rsa.dart [hunk 1]
```diff
@@ -39,8 +35,9 @@ class RSAEngine extends BaseAsymmetricBlockCipher {
   }
 
   int get outputBlockSize {
-    if (_key==null) {
-      throw new StateError("Output block size cannot be calculated until init() called");
+    if (_key == null) {
+      throw new StateError(
```
Comment (yshrsmz) [inline lib/asymmetric/rsa.dart:39]: these reformat stuff have been done by dartfmt

File: lib/key_generators/rsa_key_generator.dart [hunk 2]
```diff
@@ -7,17 +7,12 @@ library pointycastle.impl.key_generator.rsa_key_generator;
 import "package:pointycastle/api.dart";
 import "package:pointycastle/asymmetric/api.dart";
 import "package:pointycastle/key_generators/api.dart";
-import "package:pointycastle/src/registry/registry.dart";
 
 bool _testBit(BigInt i, int n) {
   return (i & (BigInt.one << n)) != BigInt.zero;
 }
 
 class RSAKeyGenerator implements KeyGenerator {
```
Comment (yshrsmz) [inline lib/key_generators/rsa_key_generator.dart:15]: this PR also includes RSA key pair generation fix provided by #135

**Comment by stevenroose** [comment]
Hmm. This goes back a lot. There's also some changes that revert more recent fixes. I think we don't necessarily have to go back to the `initCipher()` situation, where all dynamic registries are centralized. The `FACTORY_CONFIG` variables work and can be used for that instead. 

What we need is a registry that does not use reflection to receive all these `FACTORY_CONFIG` objects. It shouldn't be that hard. I think either September 1st or September 22nd I could try and implement it!

**Comment by yshrsmz** [comment]
Thanks for the quick review!

If I understand correctly, `Registry` class would take some function as a constructor parameter, which registers all related `FACTORY_CONFIG`s, right? And if so, since registry is there at a base class(such as `Digest`), does it cause circular dependency?

**Comment by stevenroose** [comment]
Hmm, a tiny bit. Logic-wise not really, just for the registration.

`Registry` needs a method `void register<T>(FactoryConfig config)`. This is fine, doesn't need any dependency on the rest of the project.

But somewhere there should be a method that registers all the algos. Preferably this can be called lazily on the first usage of the registry. Something like `Registry.ensureInit()` which calls all the `registry.register<Digest>(Md5Digest.FACTORY_CONFIG)`...

**Comment by yshrsmz** [comment]
Ah, I've misunderstood the concept. sorry.

So there's one single(possibly multiple?) global `Registry` instance, not as a static field of `Digest` or `KeyDerivator`.

**Comment by stevenroose** [comment]
Hmm, it could be type-specific, but then you will have to cross different types. F.e. "SHA-256/HMAC" will ne in the `Mac` class registry, but will have to use the `Digest` registry to construct the SHA-256 digest etc.

I could be either, I think. No actual preference or arguments for either. Maybe an argument for a single global one is that it's easier to use different implementations. F.e. a mirrors-feeded one and a hard-coded one. Or even a disabled one.

**Comment by Cretezy** [comment]
Great work! This works perfectly in Flutter (except SecureRandom, which needs to be provided but that might actually be intentional).

What other work need to be done for this PR to be put into master? I'd like to contribute as much as I can since this is quite an important PR

**Comment by yshrsmz** [comment]
Thanks!

As @stevenroose said, this PR is too out of date. So I think I'm going to close this.

But I'm now trying to implement something more similar to the current master implementation, but without reflection.

**Comment by Cretezy** [comment]
It's out of date? Your branch (`registry`) is the only thing I could get working in Flutter. What should I be using?

**Comment by yshrsmz** [comment]
I've created another PR(#148), which I believe is more sophisticated.
So closing this one.

**Pull request closed without merging**